executor/reduce_executor.rs

```diff
@@ -260,268 +232,301 @@ impl ReducerAPI for CountVarExecutor {
+                                self.min = Some(value)
+                            }
+                        } else {
+                            self.min = Some(value);
+                        }
+                    }
+                }
 
-#[derive(Debug, Clone)]
-struct MaxIntegerExecutor {
-    max: Option<i64>,
-    target: VariablePosition,
-}
+                fn finalise(self) -> Option<VariableValue<'static>> {
+                    Some(VariableValue::Value(Value::$ty(self.min?)))
+                }
+            }
 
-impl MaxIntegerExecutor {
-    fn new(target: VariablePosition) -> Self {
-        Self { max: None, target }
-    }
-}
+            #[derive(Debug, Clone)]
+            struct [< Max $ty Executor >] {
+                max: Option<$repr>,
+                target: VariablePosition,
+            }
 
-impl ReducerAPI for MaxIntegerExecutor {
-    fn accept<Snapshot: ReadableSnapshot>(
-        &mut self,
-        row: &MaybeOwnedRow<'_>,
-        context: &ExecutionContext<Snapshot>,
-        storage_counters: StorageCounters,
-    ) {
-        if let Some(value) = extract_value(row, self.target, context, storage_counters).map(|v| v.unwrap_integer()) {
-            if let Some(current) = self.max.as_ref() {
-                if value > *current {
-                    self.max = Some(value)
+            impl [< Max $ty Executor >] {
+                fn new(target: VariablePosition) -> Self {
+                    Self { max: None, target }
+                }
+            }
+
+            impl ReducerAPI for [< Max $ty Executor >] {
+                fn accept<Snapshot: ReadableSnapshot>(
+                    &mut self,
+                    row: &MaybeOwnedRow<'_>,
+                    context: &ExecutionContext<Snapshot>,
+                    storage_counters: StorageCounters,
+                ) {
+                    if let Some(value) = extract_value(row, self.target, context, storage_counters).map(|v| v.[< unwrap_ $lower >]()) {
+                        if let Some(current) = self.max.as_ref() {
+                            if value > *current {
+                                self.max = Some(value)
+                            }
+                        } else {
+                            self.max = Some(value);
+                        }
+                    }
+                }
+
+                fn finalise(self) -> Option<VariableValue<'static>> {
+                    Some(VariableValue::Value(Value::$ty(self.max?)))
                 }
-            } else {
-                self.max = Some(value);
             }
         }
-    }
+    )*};
+}
 
-    fn finalise(self) -> Option<VariableValue<'static>> {
-        self.max.map(|v| VariableValue::Value(Value::Integer(v)))
-    }
+minmax_reducer_executors! {
+    Integer::integer(i64),
+    Double::double(f64),
+    Decimal::decimal(Decimal),
+    Date::date(NaiveDate),
+    DateTime::date_time(NaiveDateTime),
+    DateTimeTZ::date_time_tz(DateTime<TimeZone>),
 }
```

Comment:
`Cow<'a, str>` is the only reason `String` is implemented separately and not in the macro. It's just enough extra handling to be more trouble than it's worth in the macro.